Multilevel Planning (bare essentials)

aorthey opened the pull request:
Our earlier pull request on multilevel motion planning was way too big, so I just cleaned it up and deleted all extensions (in particular everything about restrictions/sections (https://arxiv.org/abs/2010.14524) and about sparse multilevel roadmaps (https://arxiv.org/abs/2011.00832)).

The current pull request only contains the bare essentials which consists of all the contributions made by the paper "Multilevel Motion Planning: A Fiber Bundle Formulation" (https://arxiv.org/abs/2007.09435). The current pull request significantly extends upon an earlier pull request about quotient spaces (which was based on papers "Rapidly-Exploring Quotient-Space Trees: Motion Planning using Sequential Simplifications" (https://arxiv.org/abs/1906.01350) and "Quotient-Space Motion Planning" (https://arxiv.org/abs/1807.09468).

zkingston commented on src/ompl/multilevel/datastructures/BundleSpaceSequenceImpl.h:
```diff
@@ -0,0 +1,420 @@
+    }
+}
+
+// template <class T>
```
remove

zkingston commented on src/ompl/multilevel/datastructures/BundleSpaceSequenceImpl.h:
```diff
@@ -0,0 +1,420 @@
+            return ompl::base::PlannerStatus::TIMEOUT;
+        }
+    }
+    // double t_end = ompl::time::seconds(ompl::time::now() - t_start);
```
remove

zkingston reviewed (changes requested):
I did a quick scan through about half the code, I haven't had a chance to actually try it out on my machine yet.  I'll be back later to review it further. Some high-level comments:
- Run clang-format through all the files, there were a lot of instances of things not being according to the format.
- There are lots of leftover comments that should be removed.
- It might be good to run clang-tidy as well to keep some stylistic choices consistent between the existing OMPL code base.
- A lot of methods, their arguments, member variables, etc. are undocumented and should be.
- You don't need virtual if you also have override.

zkingston commented on src/ompl/multilevel/datastructures/importance/src/Exponential.cpp:
```diff
@@ -0,0 +1,52 @@
+
+double ompl::multilevel::BundleSpaceImportanceExponential::eval()
+{
+    const double p = 1.0 / bundleSpaceGraph_->getBundleDimension();
+    double N = (double)bundleSpaceGraph_->getNumberOfVertices();
+    double Nd = powf(N, p);
```
all const?

zkingston commented on src/ompl/multilevel/datastructures/metrics/src/BundleSpaceMetric.cpp:
```diff
@@ -0,0 +1,82 @@
+ompl::multilevel::BundleSpaceMetric::BundleSpaceMetric(BundleSpaceGraph *bundleSpaceGraph)
+  : bundleSpaceGraph_(bundleSpaceGraph)
+{
+    // if (bundleSpaceGraph_->getFiberDimension() > 0)
+    // {
+    //     xFiberStartTmp_ = bundleSpaceGraph_->getFiber()->allocState();
+    //     xFiberDestTmp_ = bundleSpaceGraph_->getFiber()->allocState();
+    // }
```
remove

zkingston commented on src/ompl/multilevel/datastructures/metrics/src/Geodesic.cpp:
```diff
@@ -0,0 +1,90 @@
+    std::cout << "NYI" << std::endl;
+    exit(0);
+    return 0.0;
+    // if (bundleSpaceGraph_->getFiberDimension() > 0)
+    // {
+    //     bundleSpaceGraph_->projectFiber(xStart->state, xFiberStartTmp_);
+    //     bundleSpaceGraph_->projectFiber(xDest->state, xFiberDestTmp_);
+    //     double d = bundleSpaceGraph_->getFiber()->distance(xFiberStartTmp_, xFiberDestTmp_);
+    //     return d;
+    // }
+    // else
+    // {
+    //     return 0.0;
+    // }
```
remove

zkingston commented on src/ompl/multilevel/datastructures/metrics/src/Geodesic.cpp:
```diff
@@ -0,0 +1,90 @@
+
+double BundleSpaceMetricGeodesic::distanceFiber(const Configuration * /*xStart*/, const Configuration * /*xDest*/)
+{
+    std::cout << "NYI" << std::endl;
```
maybe a more informative error? throw an exception? I don't like having an explicit `exit` lying around in library code.

zkingston commented on src/ompl/multilevel/datastructures/projections/FiberedProjection.h:
```diff
@@ -0,0 +1,98 @@
+        protected:
+            virtual ompl::base::StateSpacePtr computeFiberSpace() = 0;
+
+            // virtual void print(std::ostream &out) const;
```
remove

zkingston commented on src/ompl/multilevel/datastructures/projections/src/RN_RM.cpp:
```diff
@@ -0,0 +1,111 @@
+
+void Projection_RN_RM::projectFiber(const ompl::base::State *xBundle, ompl::base::State *xFiber) const
+{
+    const base::RealVectorStateSpace::StateType *xBundle_RN = xBundle->as<base::RealVectorStateSpace::StateType>();
```
General comment about many of these projections: it might be nice to use auto in a few places where there is redundant type info (e.g., here where the type is in the `as` expression).

zkingston commented on src/ompl/multilevel/datastructures/src/BundleSpace.cpp:
```diff
@@ -0,0 +1,520 @@
+        throw Exception("Space infinite measure.");
+    }
+
+    // if(space->getMeasure() <= 0)
+    // {
+    //     throw Exception("Space has zero measure.");
+    // }
```
delete

zkingston commented on src/ompl/multilevel/datastructures/src/BundleSpace.cpp:
```diff
@@ -0,0 +1,520 @@
+    BundleSpace::counter_ = 0;
+}
+
+// void BundleSpace::projectFiber(const State *xBundle, State *xFiber) const
+// {
+//     unsigned int M = components_.size();
+
+//     if (M > 1)
+//     {
+//         for (unsigned int m = 0; m < M; m++)
+//         {
+//             if (components_.at(m)->getFiberDimension() > 0)
+//             {
+//                 const State *xmBundle = xBundle->as<CompoundState>()->as<State>(m);
+//                 State *xmFiber = xFiber->as<CompoundState>()->as<State>(m);
+//                 components_.at(m)->projectFiber(xmBundle, xmFiber);
+//             }
+//         }
+//     }
+//     else
+//     {
+//         components_.front()->projectFiber(xBundle, xFiber);
+//     }
+// }
```
delete

zkingston commented on src/ompl/multilevel/datastructures/src/BundleSpace.cpp:
```diff
@@ -0,0 +1,520 @@
+}
+
+void BundleSpace::print(std::ostream &out) const
+{
```
delete this function?

zkingston commented on src/ompl/multilevel/datastructures/src/BundleSpaceGraph.cpp:
```diff
@@ -0,0 +1,980 @@
+    }
+}
+
+// void BundleSpaceGraph::setFindSectionStrategy(FindSectionType type)
+// {
+//     if (pathRestriction_ != nullptr)
+//     {
+//         pathRestriction_->setFindSectionStrategy(type);
+//     }
+// }
+// const PathRestrictionPtr BundleSpaceGraph::getPathRestriction()
+// {
+//     if (!hasBaseSpace())
+//     {
+//         OMPL_WARN("Tried getting path restriction without base space");
+//         return nullptr;
+//     }
+
+//     base::PathPtr basePath =
+//       static_cast<BundleSpaceGraph *>(getBaseBundleSpace())->getSolutionPathByReference();
+
+//     pathRestriction_->setBasePath(basePath);
+
+//     return pathRestriction_;
+// }
```
delete

zkingston replied:
same for other commented blocks throughout this file

zkingston commented on src/ompl/multilevel/datastructures/src/BundleSpaceGraph.cpp:
```diff
@@ -0,0 +1,980 @@
+    lengthStartGoalVertexPath_ = base::dInf;
+    shortestVertexPath_.clear();
+
+    qStart_ = nullptr;
+    qGoal_ = nullptr;
```
are these properly deleted?

zkingston reviewed (changes requested):
A little more commenting, I'll finish up looking through everything tomorrow.

aorthey commented on demos/multilevel/MultiLevelPlanarManipulatorCommon.h:
```diff
@@ -0,0 +1,122 @@
+        coordinates.push_back({angles[0], angles[1]});
+
+        //(1) check out of bounds
+        if (world_.outOfBounds(coordinates[0])) return false;
```
Hm weird. I actually ran clang-format over all files before commiting (using the command: find src/ompl/multilevel/ -regex '.*\.\(cpp\|hpp\|h\|cc\|cxx\)' -exec clang-format -style=file -i {} \;). It did not complain or change that part. I guess you would prefer the return on a new line?

aorthey commented on demos/multilevel/MultiLevelPlanarManipulatorCommon.h:
```diff
@@ -0,0 +1,122 @@
+        return true;
+    }
+private:
+    const PolyWorld &world_;
```
I actually just copy/pasted the structure from demos/PlanarManipulator/PlanarManipulatorStateValidityChecker.h, where it is also implemented as a reference. I guess it makes more sense to make it a pointer. I changed it now in all instances.

aorthey commented on src/ompl/base/goals/GoalState.h:
```diff
@@ -63,7 +63,7 @@ namespace ompl
 
             /** \brief Compute the distance to the goal (heuristic) */
-            double distanceGoal(const State *st) const override;
+            virtual double distanceGoal(const State *st) const override;
```
I think I implemented some custom goal state which required a new distanceGoal function, so I wanted to override distanceGoal again.

aorthey commented on src/ompl/multilevel/datastructures/BundleSpace.h:
Yes this should be ok and is used to assign unique ids to multiple bundle spaces (which is stored in the variable id_).

aorthey commented:
Hi @zkingston, thanks again for all the comments. Really appreciate your help.

I finished incorporating all your comments. Here are the biggest changes:

- Documented all methods I found lacking a documentation
- Added some missing files for computing path sections (required to make the Hypercube experiment run as expected). This was missing to make the code reproduce results from our paper (https://arxiv.org/abs/2007.09435)
- Removed all left-over comments, added auto in all projection files, checked again that all demo files work correctly
- Ran clang-format and clang-tidy on the files. Is there any agreed upon convention on which clang-tidy commands to use?

Let me know if you think there is anything else missing.

Best,
Andreas

mamoll commented:
@zkingston can you check if all you concerns have been addressed?

zkingston commented on src/ompl/tools/benchmark/src/Benchmark.cpp:
```diff
@@ -691,6 +691,7 @@ void ompl::tools::Benchmark::benchmark(const Request &req)
                     break;
             }
         }
+        planners_[i]->clear();
```
Why the clear here? Aren't the planners cleared at the top of the loop?

aorthey replied:
Yes that is true. But when I was using it to benchmark a lot of planner (like 30+), some planners did not clear their last run, which then filled up my memory. Since I did not had so much memory on my laptop, and I was fearing this would affect the results of planners which were at the very end. This clear() call just makes sure that all the temporary structures are deleted again (and nothing remains from the last run).

zkingston reviewed (changes requested):
Just want to make sure that segfault is fixed before merging. I also noticed a few formatting issues throughout, make sure to run clang-format again on every file in the code base. Also, not every member variable/method are documented, but these are for components that are deep in the multi-level planning system.

aorthey commented on demos/multilevel/MultiLevelPlanarManipulatorDemo.cpp:
```diff
@@ -0,0 +1,255 @@
+    projVec.push_back(projAB);  // Projection SE2 to X
+    siVec.push_back(si);        // State Space X
+
+    auto planner = std::make_shared<ompl::multilevel::QRRT>(siVec, projVec);
```
Yes, it must the other way round of course. I also had some other issues when specifying manual projections. Should be solved with the newest commit.

aorthey commented:
@zkingston: Do you think this looks good now, or was there anything else missing before we can merge this?

zkingston reviewed (approved):
LGTM, sorry for the delay!
